src/OpenDebugAD7/AD7DebugSession.cs

```diff
@@ -2094,7 +2154,32 @@ public void HandleIDebugEntryPointEvent2(IDebugEngine2 pEngine, IDebugProcess2 p
 
         public void HandleIDebugBreakpointEvent2(IDebugEngine2 pEngine, IDebugProcess2 pProcess, IDebugProgram2 pProgram, IDebugThread2 pThread, IDebugEvent2 pEvent)
         {
-            FireStoppedEvent(pThread, StoppedEvent.ReasonValue.Breakpoint);
+            IList<Tracepoint> tracepoints = GetTracepoints(pEvent as IDebugBreakpointEvent2);
+            if (tracepoints.Any())
+            {
+                ThreadPool.QueueUserWorkItem((o) =>
+                {
+                    foreach (var tp in tracepoints)
+                    {
+                        string logMessage = tp.GetLogMessage(pThread, Constants.EvaluationRadix, m_processName);
+
+                        m_logger.WriteLine(LoggingCategory.DebuggerStatus, logMessage);
+                    }
+                });
+            }
+
+            if (!m_isStepping && tracepoints.Any())
+            {
+                ThreadPool.QueueUserWorkItem((o) =>
+                {
+                    BeforeContinue();
+                    m_program.Continue(pThread);
+                });
+            }
+            else
+            {
+                FireStoppedEvent(pThread, StoppedEvent.ReasonValue.Breakpoint);
+            }
```

Comment:
If there were tracepoints, you want to wait for the work you scheduled to the thread pool to finish before doing either of these. I would suggest moving this into the first lambda.